doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,262 @@
+
+Any fips algorithm that is approved would then need a setter that at a minimum contains code similar to the following
+
+``` C
+int ossl_xxx_fips_approved(void)
+{
+ifdef FIPS_MODULE
+    return 1; // conditional code would go here for each algorithm if required *
+else
+    return 0;
+endif
+}
+```
+
+and in the algorithms get_ctx() function
+
+``` C
+int xxx_get_fips_approved(OSSL_PARAM params[])
+{
+    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
+        return 0;
+    return 1;
+}
+```
+
+### API’s that would be used to support this are
+
+- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
+
+``` C
+EVP_PKEY_CTX_get_params(ctx, );
+```
+
+(Note that this would mean you could not use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
+
+- Ciphers
+
+``` C
+EVP_CIPHER_CTX_get_params()
+```
+
+- Digests
+
+``` C
+EVP_MD_CTX_get_params()
+```
+
+- KDF’s
+
+``` C
+EVP_KDF_CTX_get_params()
+```
+
+- MAC’s
+
+``` C
+EVP_MAC_CTX_get_params()
+```
+
+- RAND
+
+``` C
+EVP_RAND_CTX_get_params()
+```
+
+### Backwards Compatibility
+
+Previous providers do not support this operation, so they will return not approved if they are not handled.
+
+### Alternate Solution
+
+If we had different kinds of compliance requirements (something other than FIPS) either a separate getter could be added or the getter could return a int type instead of a 0 or 1..
+(e.g 1 = fips approved, 2 = some other compliance approved)
+
+Changes Required for indicators
+-------------------------------
+
+### key size >= 112 bits
+
+There are a few places where we do not enforce key size that need to be addressed.
+
+- HMAC  Which applies to all algorithms that use HMAC also.
+- CMAC
+
+### Algorithm Transitions
+
+Should we remove these algorithms completely from the fips provider, or use indicators?
+
+- DES_EDE3_ECB.  Disallowed for encryption, allowed for legacy decryption
+- DSA.  Keygen and Signing are no longer approved, not sure if verify is still approved.
+- RSA Signing using PKCSV15.  Rsa self test for sign may need to change as the default pad_mode is PKCSV15. Check if saltlen needs a indicator. Padding mode updates required in rsa_check_padding(). Check if sha1 is allowed?
+- ECDSA B & K curves are deprecated   It looks like these are still allowed. Are the approved?
+- ED25519/ED448 is now approved.
+- X25519/X448 is not approved currently. keygen would need an indicator if we allow it?
+- RSA Key transport  Padding need to be changed to either not allow padding modes other than OEAP or use an indicator.
+- RSA - It looks like SP800-131Ar2 specifies that RSA >= 2048 is approved. Can we check with the labs what to do for the legacy verification cases since we allow >=1024. Would this now be unapproved? This would apply to rsa_kem also? rsa_keygen_pairwise_test() may need to change to do a sign/verify PCT?
```

Comment:
SP 800-131Ar2 Table 2 specifies that 1024 ≤ len(n) < 2048 is allowed for "legacy use" for signature verification purposes, so that could have an approved indicator.

For RSA KAS (RSASVE) and KTS (OAEP), SP 800-131Ar2 Table 5 does not have a "legacy use" provision. There, the minimum is 2048 bits for both encryption/encapsulation and decryption/decapsulation.

Reply:
You are right, I must of misread something.  The FIPS 140-3 IG C.F makes this clear, note that this also allows FIPS-186-2 verification also...

Reply:
For FIPS 186-2 verification, strictly speaking only "1024+256*s, where s is a nonnegative integer" would be allowed. So the approved modulus sizes would go something like: 1024, 1280, 1536, 1792, and then 2048 and above. I don't like that it's so awkward, instead of just allowing all modulus sizes in 1024-2048, but that's what they did.